File: src/Compilers/VisualBasic/Portable/Utilities/EnumerableExtensions.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic
    Module EnumerableExtensions
        <Extension>
        Public Iterator Function Concat(Of T)(source As IEnumerable(Of T), value As T) As IEnumerable(Of T)
            For Each v In source
                Yield v
            Next

            Yield value
        End Function
    End Module
End Namespace

File: src/RoslynSdk/Samples/CSharp/CSharpToVisualBasicConverter/CSharpToVisualBasicConverter.Lib/Utilities/EnumerableExtensions.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.Collections.Generic;

namespace CSharpToVisualBasicConverter.Utilities
{
    internal static class EnumerableExtensions
    {
        private static IEnumerable<T> ConcatWorker<T>(this IEnumerable<T> source, T value)
        {
            foreach (T v in source)
            {
                yield return v;
            }

            yield return value;
        }

        public static IEnumerable<T> Concat<T>(this IEnumerable<T> source, T value)
        {
            if (source == null)
            {
                throw new ArgumentNullException(nameof(source));
            }

            return source.ConcatWorker(value);
        }
    }
}
